core/src/main/scala/org/apache/spark/BarrierCoordinator.scala

```diff
@@ -173,7 +204,18 @@ private[spark] class BarrierCoordinator(
         requesters: ArrayBuffer[RpcCallContext],
         numTasks: Int): Boolean = {
       if (requesters.size == numTasks) {
-        requesters.foreach(_.reply(()))
+        if (requestMethodToSync == RequestMethod.BARRIER) {
+          requesters.foreach(_.reply(""))
+        }
+        else if (requestMethodToSync == RequestMethod.ALL_GATHER) {
+          val jsonArray = JArray(
+            allGatherMessages.map(
+              (msg) => JString(msg)
+            ).toList
+          )
+          val json: String = compact(render(jsonArray))
```

Comment:
I'm not clear about  user case, actually. But, may I ask why we change to use Json format?

Reply:
Primarily because for users using the Python API, we must deserialize the result in the python on the python side in `taskcontext.py` after it reads the result from the socket file in the `load_from_socket` function. Is there a different serialization scheme you would recommend?

Reply:
I see.